lib/livebook_web/live/session_live.ex

```diff
@@ -1641,4 +1683,25 @@ defmodule LivebookWeb.SessionLive do
        }}
     end)
   end
+
+  def assert_policy!(socket, key) do
+    unless socket.assigns.policy |> Map.fetch!(key) do
+      raise "policy not allowed"
+    end
+
+    :ok
+  end
+
+  defp assert_live_action_access!(%{assigns: %{live_action: live_action}} = socket) do
+    cond do
+      live_action in [:page, :file_settings, :runtime_settings, :bin] ->
+        assert_policy!(socket, :edit)
+
+      live_action in [:shared_page, :catch_all, :dependency_search] ->
+        assert_policy!(socket, :read)
+
+      true ->
+        raise "unknown live_action: #{inspect(live_action)}"
+    end
```

Comment:
Note for later: we will need to relax sessions controller to serve images/downloads for shared sessions.

Reply:
If we move the policy to the router and we use distinct `live_sessions`, then we do not need this. It is impossible to navigate across `live_sessions` without a full page reload, and a full page reload will always force the proper policy. So my suggestion is to remove `assert_live_action_access` and move the policy setting to the router!

Reply:
Smart, sounds good!

There is one other issue with routes. For deployments with auth proxy, people will expose only `/public`. When someone navigates from `/public/sessions/:id` to `/sessions/:id/export` and then refresh, they are no longer under `/public`, so I think the request will fail.

Reply:
@jonatanklosko we will have to expose all routes again and figure out a way of making their generation customizable based on the policy. Perhaps by passing the route helper or the live view itself as argument.

Reply:
@josevalim 
Is this right? https://github.com/livebook-dev/livebook/pull/1128/commits/550cbffeb67bc4ea51662ea1b01eda6b2a57a16f